// astro/dist/core/build/plugins/plugin-middleware.js
import { MIDDLEWARE_PATH_SEGMENT_NAME } from "../../constants.js";
import { addRollupInput } from "../add-rollup-input.js";
const MIDDLEWARE_MODULE_ID = "@astro-middleware";
const RESOLVED_MIDDLEWARE_MODULE_ID = "\0@astro-middleware";
let inputs = /* @__PURE__ */ new Set();
function vitePluginMiddleware(opts, _internals) {
  return {
    name: "@astro/plugin-middleware",
    options(options) {
      if (opts.settings.config.experimental.middleware) {
        return addRollupInput(options, [MIDDLEWARE_MODULE_ID]);
      }
    },
    resolveId(id) {
      if (id === MIDDLEWARE_MODULE_ID && opts.settings.config.experimental.middleware) {
        return RESOLVED_MIDDLEWARE_MODULE_ID;
      }
    },
    async load(id) {
      if (id === RESOLVED_MIDDLEWARE_MODULE_ID && opts.settings.config.experimental.middleware) {
        const imports = [];
        const exports = [];
        let middlewareId = await this.resolve(
          `${opts.settings.config.srcDir.pathname}/${MIDDLEWARE_PATH_SEGMENT_NAME}`
        );
        if (middlewareId) {
          imports.push(`import { onRequest } from "${middlewareId.id}"`);
          exports.push(`export { onRequest }`);
        }
        const result = [imports.join("\n"), exports.join("\n")];
        return result.join("\n");
      }
    }
  };
}
function pluginMiddleware(opts, internals) {
  return {
    build: "ssr",
    hooks: {
      "build:before": () => {
        return {
          vitePlugin: vitePluginMiddleware(opts, internals)
        };
      }
    }
  };
}
export {
  MIDDLEWARE_MODULE_ID,
  RESOLVED_MIDDLEWARE_MODULE_ID,
  pluginMiddleware,
  vitePluginMiddleware
};
